solidity/contracts/middleware/liquidity-layer/adapters/PortalAdapter.sol

```diff
@@ -0,0 +1,221 @@
+
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+contract PortalAdapter is ILiquidityLayerAdapter, Router {
+    /// @notice The Portal TokenBridge contract.
+    IPortalTokenBridge public portalTokenBridge;
+
+    /// @notice The LiquidityLayerRouter contract.
+    address public liquidityLayerRouter;
+
+    /// @notice Hyperlane domain => Wormhole domain.
+    mapping(uint32 => uint16) public hyperlaneDomainToWormholeDomain;
+    /// @notice transferId => token address
+    mapping(bytes32 => address) public portalTransfersProcessed;
+
+    uint32 localDomain;
+    // We could technically use Portal's sequence number here but it doesn't
+    // get passed through, so we would have to parse the VAA twice
+    uint224 public nonce = 0;
+
+    /**
+     * @notice Emits the nonce of the Portal message when a token is bridged.
+     * @param nonce The nonce of the Portal message.
+     * @param portalSequence The sequence of the Portal message.
+     * @param destination The hyperlane domain of the destination
+     */
+    event BridgedToken(
+        uint256 nonce,
+        uint64 portalSequence,
+        uint32 destination
+    );
+
+    /**
+     * @notice Emitted when the Hyperlane domain to Wormhole domain mapping is updated.
+     * @param hyperlaneDomain The Hyperlane domain.
+     * @param wormholeDomain The Wormhole domain.
+     */
+    event DomainAdded(uint32 indexed hyperlaneDomain, uint32 wormholeDomain);
+
+    modifier onlyLiquidityLayerRouter() {
+        require(msg.sender == liquidityLayerRouter, "!liquidityLayerRouter");
+        _;
+    }
+
+    /**
+     * @param _localDomain The local hyperlane domain
+     * @param _owner The new owner.
+     * @param _portalTokenBridge The Portal TokenBridge contract.
+     * @param _liquidityLayerRouter The LiquidityLayerRouter contract.
+     */
+    function initialize(
+        uint32 _localDomain,
+        address _owner,
+        address _portalTokenBridge,
+        address _liquidityLayerRouter
+    ) public initializer {
+        // Transfer ownership of the contract to deployer
+        _transferOwnership(_owner);
+
+        // Set the addresses for the ACM and IGP to address(0) - they aren't used.
+        _setAbacusConnectionManager(address(0));
+        _setInterchainGasPaymaster(address(0));
+
+        localDomain = _localDomain;
+        portalTokenBridge = IPortalTokenBridge(_portalTokenBridge);
+        liquidityLayerRouter = _liquidityLayerRouter;
+    }
+
+    /**
+     * Sends tokens as requested by the router
+     * @param _destinationDomain The hyperlane domain of the destination
+     * @param _token The token address
+     * @param _amount The amount of tokens to send
+     */
+    function sendTokens(
+        uint32 _destinationDomain,
+        bytes32, // _recipientAddress, unused
+        address _token,
+        uint256 _amount
+    ) external onlyLiquidityLayerRouter returns (bytes memory) {
```

Comment:
do you not need `override` for here and `receiveTokens`? because of this `is ILiquidityLayerAdapter`

Reply:
Good question, not sure why. Certainly compiles without

Reply:
I thought so too but apparently not